packages/auth/test/integration/flows/totp.test.ts

```diff
@@ -0,0 +1,156 @@
+import { FirebaseError } from '@firebase/app';
+import {
+  cleanUpTestInstance,
+  getTestInstance,
+  getTotpCode,
+  delay,
+  email,
+  incorrectTotpCode
+} from '../../helpers/integration/helpers';
+
+import {
+  TotpMultiFactorGenerator,
+  TotpSecret
+} from '../../../src/mfa/assertions/totp';
+
+use(chaiAsPromised);
+use(sinonChai);
+
+describe(' Integration tests: Mfa TOTP', () => {
+  let auth: Auth;
+  let totpSecret: TotpSecret;
+  let displayName: string;
+  beforeEach(async () => {
+    auth = getTestInstance();
+    displayName = 'totp-integration-test';
+  });
+
+  afterEach(async () => {
+    await cleanUpTestInstance(auth);
+  });
+
+  it('should not enroll if incorrect totp supplied', async () => {
+    const cr = await signInWithEmailAndPassword(auth, email, 'password');
+    const mfaUser = multiFactor(cr.user);
+    const session = await mfaUser.getSession();
+    totpSecret = await TotpMultiFactorGenerator.generateSecret(session);
+    const multiFactorAssertion =
+      TotpMultiFactorGenerator.assertionForEnrollment(
+        totpSecret,
+        incorrectTotpCode
+      );
+
+    await expect(
+      mfaUser.enroll(multiFactorAssertion, displayName)
+    ).to.be.rejectedWith('auth/invalid-verification-code');
+  });
+
+  it('should enroll using correct otp', async () => {
+    const cr = await signInWithEmailAndPassword(auth, email, 'password');
+
+    const mfaUser = multiFactor(cr.user);
+
+    const session = await mfaUser.getSession();
+
+    totpSecret = await TotpMultiFactorGenerator.generateSecret(session);
+
+    const totpVerificationCode = getTotpCode(
+      totpSecret.secretKey,
+      totpSecret.codeIntervalSeconds,
+      totpSecret.codeLength
+    );
+
+    const multiFactorAssertion =
+      TotpMultiFactorGenerator.assertionForEnrollment(
+        totpSecret,
+        totpVerificationCode
+      );
+    await expect(mfaUser.enroll(multiFactorAssertion, displayName)).to.be
+      .fulfilled;
+  });
+
+  it('should not allow sign-in with incorrect totp', async () => {
+    let resolver;
+
+    try {
+      await signInWithEmailAndPassword(auth, email, 'password');
+
+      throw new Error('Signin should not have been successful');
+    } catch (error) {
+      expect(error).to.be.an.instanceOf(FirebaseError);
+      expect((error as any).code).to.eql('auth/multi-factor-auth-required');
+
+      resolver = getMultiFactorResolver(auth, error as any);
+      expect(resolver.hints).to.have.length(1);
+
+      const assertion = TotpMultiFactorGenerator.assertionForSignIn(
+        resolver.hints[0].uid,
+        incorrectTotpCode
+      );
+
+      await expect(resolver.resolveSignIn(assertion)).to.be.rejectedWith(
+        'auth/invalid-verification-code'
+      );
+    }
+  });
+
+  it('should allow sign-in with for correct totp and unenroll successfully', async () => {
+    let resolver;
+
+    await delay(30 * 1000);
+    // Added a delay so that getTotpCode() actually generates a new totp code
```

Comment:
add a "// TODO(bhparijat) generate the otp code for the next time window by passing the appropriate timestamp to avoid the 30s delay. The delay is needed because the otp code used for enrollment cannot be reused for signing in.